packages/datadog-instrumentations/src/redis.js

```diff
@@ -149,24 +134,28 @@ addHook({ name: 'redis', versions: ['>=0.12 <2.6'] }, redis => {
   return redis
 })
 
-function start (client, command, args, url = {}) {
-  const db = client.selected_db
-  const connectionOptions = client.connection_options || client.connection_option || client.connectionOption || url
-  startCh.publish({ db, command, args, connectionOptions })
+function getStartCtx (client, command, args, url = {}) {
+  return {
+    db: client.selected_db,
+    command,
+    args,
+    connectionOptions: client.connection_options || client.connection_option || client.connectionOption || url
+  }
 }
 
-function wrapCallback (finishCh, errorCh, callback) {
+function wrapCallback (finishCh, errorCh, ctx, callback) {
   return shimmer.wrapFunction(callback, callback => function (err) {
-    finish(finishCh, errorCh, err)
-    if (callback) {
-      return callback.apply(this, arguments)
-    }
+    return finish(finishCh, errorCh, ctx, err, callback, this, arguments)
   })
 }
 
-function finish (finishCh, errorCh, error) {
+function finish (finishCh, errorCh, ctx, error, callback, thisArg, args) {
   if (error) {
-    errorCh.publish(error)
+    ctx.error = error
+    errorCh.publish(ctx)
+  }
+  if (callback) {
+    return finishCh.runStores(ctx, callback, thisArg, ...args)
   }
-  finishCh.publish()
+  finishCh.publish(ctx)
```

Comment:
This will call `publish` a second time since the previous line also calls `publish` through `runStores`.

Reply:
Oh, you're right I totally missed that this only happens when there is no callback. Disregard the above.